File: src/main/java/fun/bookish/vertx/auth/simple/core/Subject.java
package fun.bookish.vertx.auth.simple.core;

import fun.bookish.vertx.auth.simple.configurable.RememberMePersistStrategy;
import fun.bookish.vertx.auth.simple.configurable.SessionIdStrategy;
import fun.bookish.vertx.auth.simple.configurable.SessionPersistStrategy;
import fun.bookish.vertx.auth.simple.constant.SimpleAuthConstants;
import fun.bookish.vertx.auth.simple.provider.SimpleAuthProvider;
import io.vertx.core.AsyncResult;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.Vertx;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.auth.User;
import io.vertx.ext.web.Cookie;
import io.vertx.ext.web.RoutingContext;

import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;


public class Subject {

    private String sessionId;
    private Vertx vertx;
    private SimpleAuthProvider authProvider;
    private SessionPersistStrategy sessionPersistStrategy;
    private SessionIdStrategy sessionIdStrategy;
    private RememberMePersistStrategy rememberMePersistStrategy;
    private Long rememberMeTimeout;

    public Subject(){}

    public Subject(String sessionID, Vertx vertx, SimpleAuthProvider authProvider, SimpleAuthOptions options){
        this.sessionId = sessionID;
        this.vertx = vertx;
        this.authProvider = authProvider;
        this.sessionPersistStrategy = options.getSessionPersistStrategy();
        this.sessionIdStrategy = options.getSessionIdStrategy();
        this.rememberMePersistStrategy = options.getRememberMePersistStrategy();
        this.rememberMeTimeout = options.getRememberMeTimeout();
    }

    private AtomicReference<User> authUserRef = new AtomicReference<>();

    private AtomicBoolean rememberMeRef = new AtomicBoolean(false);

    /**
     * 用户登录
     */
    public void login(RoutingContext ctx, JsonObject authInfo, Handler<AsyncResult<Void>> resultHandler){
        this.vertx.executeBlocking(future -> authProvider.authenticate(authInfo, res -> {
            if(res.succeeded()){
                this.authUserRef.set(res.result());
                ctx.session().put(SimpleAuthConstants.SUBJECT_KEY_IN_SESSION,this);
                try{
                    //持久化session
                    sessionPersistStrategy.cache(ctx.session());
                    //判断是否需要rememberMe
                    if(this.rememberMeRef.get()){
                        Cookie cookie = Cookie.cookie("RememberMe",ctx.session().id())
                                                .setMaxAge(rememberMeTimeout)
                                                .setHttpOnly(false).setPath("/");
                        rememberMePersistStrategy.cache(cookie,ctx.session());
                        ctx.addCookie(cookie);
                    }
                }catch (Exception e){
                    future.fail(e);
                }
                future.complete();
            }else{
                future.fail(res.cause());
            }
        }),resultHandler);
    }

    /**
     * 验证用户权限
     */
    public void isAuthorised(String authority, Handler<AsyncResult<Boolean>> resultHandler){
        if(this.authUserRef.get() == null){
            resultHandler.handle(Future.succeededFuture(false));
        }else{
            authUserRef.get().isAuthorised(authority, resultHandler);
        }
    }

    /**
     * 用户注销
     */
    public void logout(RoutingContext ctx){

        this.authUserRef.set(null);
        ctx.setUser(null);
        sessionPersistStrategy.remove(ctx.session());
        Cookie rememberMeCookie = ctx.removeCookie("RememberMe");
        if(rememberMeCookie != null){
            rememberMePersistStrategy.remove(rememberMeCookie);
        }

    }

    public boolean isAuthenticated(){
        return this.authUserRef.get() != null;
    }

    public User getUser(){
        return this.authUserRef.get();
    }

    public JsonObject getPrincipal() {
        return this.authUserRef.get()==null? null : this.authUserRef.get().principal();
    }

    public Subject enableRememberMe(boolean state){
        this.rememberMeRef.set(state);
        return this;
    }

}

File: src/main/kotlin/cn/booklish/vertx/simple/core/Subject.kt
package cn.booklish.vertx.simple.core

import cn.booklish.vertx.simple.manager.SecurityManager
import io.vertx.core.AsyncResult
import io.vertx.core.Future
import io.vertx.core.Handler
import io.vertx.core.json.JsonObject
import io.vertx.ext.auth.AuthProvider
import io.vertx.ext.auth.User
import io.vertx.ext.web.Cookie
import io.vertx.ext.web.RoutingContext
import java.util.*
import java.util.concurrent.atomic.AtomicReference


class Subject(private val authProvider: AuthProvider) {

    private var authUser: AtomicReference<User> = AtomicReference()

    private val rememberMe:AtomicReference<Boolean> = AtomicReference(false)

    fun isAuthenticated():Boolean {
        return authUser.get() != null
    }

    private fun setUser(user:User){
        this.authUser.compareAndSet(null,user)
    }

    fun getUser():User?{
        return this.authUser.get()
    }

    fun login(ctx: RoutingContext, authInfo:JsonObject, resultHandler:Handler<AsyncResult<Void>>){
        authProvider.authenticate(authInfo){
            if(it.succeeded()){
                this.setUser(it.result())

                //判断rememberMe
                val rememberMe = authInfo.getValue("RememberMe")
                if(rememberMe != null){
                    val booleanType = (rememberMe is Boolean) && rememberMe
                    val stringType = (rememberMe is String) && rememberMe.equals("true",true)
                    if(booleanType || stringType){
                        this.rememberMe.compareAndSet(false,true)
                        val cookieValue = Base64.getEncoder().encodeToString(authInfo.getString("username").toByteArray(Charsets.UTF_8))
                        val cookie = Cookie.cookie("simple-auth.RememberMe",cookieValue).setHttpOnly(false).setPath("/")
                        ctx.addCookie(cookie)
                        SecurityManager.cacheRememberMeSubject(cookieValue,this)
                    }
                }

                resultHandler.handle(Future.succeededFuture())
            }else{
                resultHandler.handle(Future.failedFuture(it.cause()))
            }
        }
    }

    fun logout(){
        this.authUser.set(null)
        this.rememberMe.set(false)
    }

    fun isAuthorised(authority:String, resultHandler:Handler<AsyncResult<Boolean>>){
        val authUser = this.authUser.get()
        if(authUser == null){
            resultHandler.handle(Future.succeededFuture(false))
        }else{
            authUser.isAuthorised(authority, resultHandler)
        }
    }

    fun isRememberMe():Boolean{
        return this.rememberMe.get()
    }

    fun getPrincipal():JsonObject{
        return this.getUser()?.principal()?: JsonObject()
    }

}